### ISSUE TITLE
create galaxy integration doc / example

### ISSUE THREAD

**Issue by nathandunn** [GMOD/Apollo#1016]
- [x] get a working docker-webapollo2 instance for 1.7+
  - [x] fix JBrowse build off of image
  - [x] upgrade apollo2 version to 2.0.2
  - [x] remove un-needed files (e.g., sdk)
  - [x] clean out docker image (wasn't much to remove)
  -  ~~add global volumes (/opt/apollo/)~~ - not well supported for the mac, but added instructions
- [x] create GMOD/docker-apollo:
  - [x] merge into [GMOD version](https://github.com/GMOD/docker-apollo)
  - [x] merge in the docker directions from #1025  into master 
- [x] add apollo bibtex to tool
  - [x] publish apollo tool

===

- [x] get [galaxy recipe](https://github.com/erasche/docker-recipes/tree/master/galaxy-apollo) working from [this documentation](https://github.com/TAMU-CPT/galaxy-webapollo)
- [x] finish out [todo list](https://github.com/GMOD/Apollo/issues/1016#issuecomment-217011731)
- [x] update  [documentation](https://github.com/GMOD/Apollo/blob/master/docs/Setup.md) to reflect galaxy docker or add a separate area

**Comment by nathandunn** [comment]
can install as a PIP file . . . will record directions here:

https://github.com/erasche/python-apollo

**Comment by nathandunn** [comment]
Steps (so far):

1. [Get galaxy](https://wiki.galaxyproject.org/Admin/GetGalaxy)
1. Create docker image with Apollo? 
1. Install docker image into Python?

**Comment by erasche** [comment]
Given that there are docker images for both Galaxy and apollo, we could probably simplify this process a lot by using `docker-compose`

We can base an apollo specific docker image off of the official galaxy image https://github.com/bgruening/docker-galaxy-stable (maintained in one of the team member's private repos), where we can simply add a couple lines to install my Apollo tools.

As part of the docker-compose.yml one can easily spin up an apollo instance as well, and network them together.

I can throw together a quick POC tomorrow if you're interesting.

**Comment by nathandunn** [comment]
@erasche   Yes a quick POC would be great.    Thanks.    

Let me know what the NeatHTMLFeatures issue is.  We just added some additional configuration for the NeatHTMLFeatures track that allows custom configuration (i.e. turning off) of certain components in the JBrowse master.

**Comment by erasche** [comment]
Making progress here

https://github.com/erasche/docker-recipes/tree/master/galaxy-apollo

what's left:

- need to get jbrowse tools to show up
- need to point apollo tools at apollo instance
- need to set up a proxy in front which allows access to both galaxy + apollo over same URL. This will be required so iframes / etc will work properly.

**Comment by nathandunn** [comment]
Thanks.   I was planning on diving deeply into this tomorrow.  Is there a place you would recommend me starting that would be the most helpful?    My end-goals are:

1 - have a galaxy server with Apollo running in it that allows for some means of getting data in and out of Apollo (and expand web-services to potentially facilitate this)
2 - provide easy documentation to do this for those running galaxy
3 - (a nice to have) provide an Apollo docker container with instructions

Very much appreciate you work on this.  

We did something similar with ansible using 2.0.2:

https://github.com/monarch-initiative/monarch-devops/blob/master/ansible/ohsu-migration/apollo.yml
https://github.com/monarch-initiative/monarch-devops/tree/master/ansible/ohsu-migration/roles/apollo

which are actively running.

**Comment by erasche** [comment]
I'm definitely on board with all of these end goals. We have grant funding which somewhat covers this topic (packaging up an Apollo + Galaxy + Chado environment as a deployable docker unit that's easily installable/available. It's a high priority for us.)

> 1 - have a galaxy server with Apollo running in it that allows for some means of getting data in and out of Apollo (and expand web-services to potentially facilitate this)

This is half-way there as-is.

This summer I'm going to work on adding OAuth to Apollo and Galaxy which is the main thing preventing deeper integartion (e.g. "blast this seq on my local galaxy with our special blast DB"). That'll include adding API key based API authentication, as well as more "RESTful" API routes, assuming I can manage that. (I'm still pretty rubbish at groovy/grails.)

The only concern I had, which I wanted to discuss with y'all at some point, was the best way to get JBrowse instances in and out of Apollo.

I do not love that we have to have Galaxy and Apollo sharing a filesystem where JBrowse instances are stored. This prevents me from generalizing the Galaxy tools to be something which can talk to *any* Apollo instance (as the user, of course, not via the current "proxy user" scenario)

Do you think it's possible that there could be an API on the Apollo side which could build up the JBrowse instance, much like my current Galaxy tool does? Here is a very rough sketch of what the endpoints might look like:

```
POST     /organism                      - create new organism, returns an organism ID for future queries
POST     /organism/:id/reference        - adds a new reference sequence to the genome
GET      /organism/:id/reference        - fetch a list of reference sequences
GET      /organism/:id/reference/:refId - fetch fasta sequence for that reference sequence
POST     /organism/:id/tracks           - add a new track. Would take a JSON object with track metadata/configuration
POST/PUT /organism/:id/tracks/:trackId  - would allow uploading data to that track. If you defined it as a BAM in the above call, you would then use this URL for uploading a BAM dataset.

```

If:

- we had an API like the above
- users could create + manage their own organisms

If there's an API and users have those sorts of permissions, then we could develop something like the current JBrowse tool, except within Apollo, and let users upload + configure tracks. 

The Galaxy tools could be expanded to create organism + configure the org's JBrowse instances for remote servers which would be quite useful for the Apollo community. 



> 2 - provide easy documentation to do this for those running galaxy

Yes, I've been meaning to package up these apollo tools for a while in a better documented package. I'm not sure what all needs to be done here (still swamped with grading/etc until the end of the semester), but if you have action items then great!

Once I get remote services implemented in Galaxy, and there's a way to add data over an API, the admin side instructions should be as simple as "Click the install apollo button, make sure you're on a recent Galaxy release."

The end user side instructions need a lot of help.

> 3 - (a nice to have) provide an Apollo docker container with instructions

Nice to have? This is requirement ;) 

My containers have been working well in production for us. Is there anything else you think needs to be done here? I'm happy to upstream it but right now it needs changes on a per-release basis. I'm assuming this will stabilise in the future.

> We did something similar with ansible using 2.0.2:

Hey, ansible! That looks great. I know the galaxy guys sure love it for deploying. I... do not love the all-in-one containers, otherwise I'd switch immediately to using that as the base Apollo container.

**Comment by nathandunn** [comment]
>> 3 - (a nice to have) provide an Apollo docker container with instructions
Nice to have? This is requirement ;)

> My containers have been working well in production for us. Is there anything else you think needs to be done here?

I just want to add some doc for this.  Could be something very simple like: "(1) install docker (link)", (2) "docker start erasch/webapollo2" (3) connect at <host>:8080/apollo  .  etc.  I can take care of that though.  

WRT to the organism management, @childers  team at the NAL is doing something similar I believe using the API.   I'll try to comment on the individual queries in just a second.  

tldr; we either already have them or we had planned to add them

**Comment by erasche** [comment]
@nathandunn great to hear!

**Comment by nathandunn** [comment]
Adding / updating organisms is here:

You can use the UrlMapper to remap existing URL's: https://github.com/GMOD/Apollo/blob/master/grails-app/conf/UrlMappings.groovy

http://icebox.lbl.gov/Apollo2/jbrowse/web_services/api

```
POST     /organism                      - create new organism, returns an organism ID for future queries
```

[Support here]: (https://github.com/GMOD/Apollo/blob/master/grails-app/controllers/org/bbop/apollo/OrganismController.groovy#L129
)

```
POST     /organism/:id/reference        - adds a new reference sequence to the genome
```
I would problem do this in the [SequenceController](https://github.com/GMOD/Apollo/blob/master/grails-app/controllers/org/bbop/apollo/SequenceController.groovy).  You will have to remove the ["read-only" cache](https://github.com/GMOD/Apollo/blob/master/grails-app/domain/org/bbop/apollo/Sequence.groovy#L22), but using this URL is fine.  

```
GET      /organism/:id/reference        - fetch a list of reference sequences
```
Same as above, should be easy to do . . same note as above.

```
GET      /organism/:id/reference/:refId - fetch fasta sequence for that reference sequence
```
Already in the web services as well (in IOServices) as a web services. 


You'll have to create a new table for this (probably storing the JSON String directly).  The trick part is loading them dynamically into the tracks table when it gets imported.   This is the thing that will require the most work, but it should be very doable.

```
POST     /organism/:id/tracks           - add a new track. Would take a JSON object with track metadata/configuration
POST/PUT /organism/:id/tracks/:trackId  - would allow uploading data to that track. If you defined it as a BAM in the above call, you would then use this URL for uploading a BAM dataset.
```

**Comment by nathandunn** [comment]
> I do not love that we have to have Galaxy and Apollo sharing a filesystem where JBrowse instances are stored. This prevents me from generalizing the Galaxy tools to be something which can talk to any Apollo instance (as the user, of course, not via the current "proxy user" scenario)

@erasche  (see note above if you didn't get it)

Let me know how else you might want to do this.  We have a great deal of flexibility with how we handle this.  

Some of the other stuff I would be happy to stub out for you to get you going, so let me know when you're ready to start it.

**Comment by nathandunn** [comment]
@erasche   A couple of notes.  

One one of my machines I am getting this error:

```WARNING: The DB_1_PORT_5432_TCP_ADDR variable is not set. Defaulting to a blank string.```

any ideas on why its not set, or any way to explicitly set (or set a sane default)?  

Also, it looks like the galaxy-apollo is trying to install jbrowse, which is fine.  If you're doing it for Apollo, its un-necessary, though as Apollo has to install its own jbrowse to properly configure it.

**Comment by erasche** [comment]
@nathandunn it's due to https://github.com/erasche/docker-recipes/blob/master/galaxy-apollo/docker-compose.yml#L19

This is a problem between different versions of docker-compose. What version are you on? I'm on 1.6.2, but it's been a few days since I've booted the image so naturally I've forgotten if I was seeing that error message or not.

Originally you could just stick env variables in docker-compose and everything was happy. Lately you've had to escape them, since I don't want *my* environment variables, I want the container's environment variable.

The "install Jbrowse" portion is to install the JBrowse tool in Galaxy, which is, of course, necessary if you want to build up any JBrowse instances and deploy them to the attached Apollo image.

**Comment by nathandunn** [comment]
@erasche   I'm glad I asked.  I have one machine with docker-compose 1.7 installed that does NOT work (the error above) and one with 1.4 which works great.   

Looks like I had an alternate version installed.    I'm trying to figure out what mixture of versions might work best and if  I should remove all of the homebrew stuff or use Docker Toolbox directly . 

If it continues to be a problem, I think that I can set DB_1_PORT_5432_TCP_ADDR to localhost and it should work.

**Comment by erasche** [comment]
Ok, I'll see if I can figure out the fix for later versions of docker-compose.

`DB_1_PORT_5432_TCP_ADDR` cannot be set to localhost, unfortunately. The value of that environment variable is populated at boot by the docker-compose command, providing the container with an IP address at which the `DB` container is reachable. This will change between launches of the `docker-compose.yml` file, and if containers are spread across infrastructure.

Having it in `docker-compose.yml` isn't optimal, it is generally considered better practice to have it inside some .sh script inside the container so the user doesn't have to think about it / deal with it, since it isn't *actually* a knob they can turn.

**Comment by nathandunn** [comment]
Yeah, I upgraded to 1.7 on the other machine using the default docker toolbox and it now consistently errors out. 

That being said, there is a python and a rails example:

https://docs.docker.com/compose/rails/ <https://docs.docker.com/compose/rails/>
https://docs.docker.com/compose/django/ <https://docs.docker.com/compose/django/>

In both cases they set host to ‘db'.  Which, seems to be connecting to a database, though I’m running into other issues, so take it with a grain of salt. 

Nathan

**Comment by erasche** [comment]
**Oh Wow**.

That's really awesome, I had not seen that. They've finally fixed this issue. Ok, I'll update my scripts for the new docker-compose.yml soon.

**Comment by nathandunn** [comment]
I am not entirely sure if it works all the way, but it seems to be connecting to the database.  Still not able to view externally, but that could be due to something else.  

Nathan

**Comment by nathandunn** [comment]
Everything else works, but the JBrowse doesn't seem to be built properly (src/dojo/dojo.js is not included) in the webapollo2 image I ran with ```docker run erasche/webapollo2:cpt```.

**Comment by nathandunn** [comment]
I just need to make a few changes to the Dockerfile and rebuild.

**Comment by nathandunn** [comment]
FYI @enuggetry  @erasche @jnguyenx    Looks like the older version of docker-compose doesn't support volumes?  Though somehow we mapped the /data file internally.   Not sure.  

https://docs.docker.com/compose/compose-file/

To upgrade:
https://docs.docker.com/compose/compose-file/#versioning

https://docs.docker.com/compose/compose-file/#upgrading

**Comment by nathandunn** [comment]
Turns out that this is badly supported on the mac:  http://stackoverflow.com/questions/23439126/how-to-mount-host-directory-in-docker-container

For now I'll just punt and use the absolute directory.

**Comment by nathandunn** [comment]
I mean use the relative directory

**Comment by nathandunn** [comment]
think that this is done now
